### DISCUSSION TITLE
Can we run PyTest on are packages with this?

### DISCUSSION THREAD

**Discussion by dreamer-coding** [mesonbuild/meson-python#635] [Q&A] [answered]
Is it possible to write test cases in Python for a Python extension written language XYZ using something like PyTest to validate the functionality in the extensions using Meson-Python layer?

**Comment by rgommers** [comment, marked as answer]
Sure. This question isn't very related to `meson-python`. `meson-python` lets you install Python packages built with Meson. If that package has `pytest`-compatible tests, then yes you can run `pytest`. 

The language that a package is written in shouldn't matter here, as long as the package builds fine. E.g., SciPy has extension modules in C, C++, Cython and Fortran. They all have tests written in Python.

**Reply by dreamer-coding** [reply]
So is possible to say write an module for fscl-xcore-c as fscl-xcore-py and write test cases using PyTest without worrying about the underlining details?

**Reply by rgommers** [reply]
Generally, yes. But again, the "create Python bindings for the C code is unrelated to `meson-python`. You'll need to create such bindings with the Python C API (see https://docs.python.org/3/extending/index.html), possibly with the use of a tool like Cython.

Once you have that, you can write `pytest` tests, and `meson-python` can be used to package/install.

**Reply by rgommers** [reply]
Here is a toy example using the Python C API: https://github.com/mesonbuild/meson-python/tree/main/tests/packages/purelib-and-platlib

And here is one that also uses Cython (the `.pyx` files): https://github.com/mesonbuild/meson-python/tree/main/tests/packages/complex

**Reply by dreamer-coding** [reply]
How about this let's assume I use this to write a Python extinction and the libraries I'm writing this for happen to use wrap files. Does this work?

**Reply by dnicolodi** [reply]
How about you read the documentation, try it out, and ask specific questions when you encounter a problem?

**Reply by dreamer-coding** [reply]
Sure, I just happened to have the perfect project to try it out for myself.
